### PR TITLE
env: Add surrogate for pytest.deprecated_call for ptyest<3.9

### LINKED ISSUE #2922: [BUG] test_builtin_caster.test_int_convert: Fails on certain minor version of CPython 3.8?
## Issue description

Test fails on current mainline branch:
https://github.com/pybind/pybind11/blob/0e01c243c7ffae3a2e52f998bacfe82f56aa96d9/tests/test_builtin_casters.py#L302-L305

## Reproducible example code

```sh
$ cd pybind11
$ git log -n 1 --oneline --no-decorate
0e01c243 Generalizing suppression for pypocketfft. (#2896)
$ mkdir build && cd build
$ cmake .. -GNinja -DPYBIND11_PYTHON_VERSION=3.8 -DPYBIND11_TEST_OVERRIDE=test_builtin_casters.cpp
...
-- PYTHON 3.8.0
...
$ env PYTEST_ADDOPTS="-s -x" ninja pytest
...
        # TODO: Avoid DeprecationWarning in `PyLong_AsLong` (and similar)
        if (3, 8) <= env.PY < (3, 10):
            with pytest.deprecated_call():
>               assert convert(Int()) == 42
E               AssertionError: Did not produce DeprecationWarning or PendingDeprecationWarning
```

### PR DESCRIPTION
**Pull request by EricCousineau-TRI**
## Description

Resolves #2922

## Suggested changelog entry:

Testing only; doesn't merit a change?

### CODE REVIEW

**Comment by EricCousineau-TRI** [comment]
@YannickJadoul or @rwgk Any chance y'all have time for a quick stamp?

**Comment by EricCousineau-TRI** [comment]
I'd also be fine with telling pybind11 to fail fast if an old version of `pytest` is found (e.g. require >=3.9).

Lemme know!

**Comment by rwgk** [comment]
> I'd also be fine with telling pybind11 to fail fast if an old version of `pytest` is found (e.g. require >=3.9).
> 
> Lemme know!

I just looked, pytest 3.9 was released in October 2018, nearly 2.5 years ago.
My vote: `require >=3.9`
In the interest of simplicity/clarity/long-term maintenance.

**Comment by EricCousineau-TRI** [comment]
Er, only caveat is that Ubuntu 18.04 also provisions pytest==3.3.2 for CPython 3.6.9 :(

```
$ dpkg -S $(which python3.6)
python3.6-minimal: /usr/bin/python3.6
$ python3.6 -m pytest --version
This is pytest version 3.3.2, imported from /usr/lib/python3/dist-packages/pytest.py
```

In that case, it'd be *requisite* that testing be done inside a `venv` per the `CONTRIBUTING` instructions.

We OK making that constraint? (and do we do that in CI?)

**Comment by EricCousineau-TRI** [comment]
Reviewing CI, seems like yeah, `venv` is always used. Kewl, will now constrain.

**Comment by henryiii** [comment]
I like this solution; unlike a normal Python package, we are not able to require Python things; so being able to use pytest in the default environment for reasonably recent OS's (Ubuntu 18.04) is nice.

**Review by henryiii — APPROVED** [review]

**Comment by henryiii** [comment]
I think we used to use the default pytest, and this got changed in the CI to force a newer pytest. As long as it's clearly marked to be removed when we do bump the minimum pytest, I think it's fine. This is useful for deploying pybind11 in package managers, where they generally use the system packages for dependencies.

**Review by rwgk — APPROVED** [review]
In support of Henry's approval.

**Comment by EricCousineau-TRI** [comment]
Gotcha, thank ya for the rationale and and history - and happy to merge in with less work haha.

File: tests/env.py [hunk 1]
```diff
@@ -12,3 +14,19 @@
...
+
+def deprecated_call():
+    """
+    pytest.deprecated_call() seems broken in pytest<3.9.x; concretely, it
+    doesn't work on CPython 3.8.0 with pytest==3.3.2 on Ubuntu 18.04 (#2922).
+
+    This is a narrowed reimplementation of the following PR :(
+    https://github.com/pytest-dev/pytest/pull/4104
+    """
+    pieces = pytest.__version__.split(".")
```
Comment (EricCousineau-TRI) [inline tests/env.py:27]: Working: Will add explicit comment to remove this when min. `pytest` version is required.

Comment (EricCousineau-TRI) [reply tests/env.py:28]: Done 7cd83686

**Pull request merged**